src/vs/base/browser/ui/iconLabel/iconLabel.ts

```diff
@@ -155,18 +162,74 @@ export class IconLabel extends Disposable {
 
 			if (this.descriptionNode instanceof HighlightedLabel) {
 				this.descriptionNode.set(description || '', options ? options.descriptionMatches : undefined);
-				if (options?.descriptionTitle) {
-					this.descriptionNode.element.title = options.descriptionTitle;
-				} else {
-					this.descriptionNode.element.removeAttribute('title');
-				}
+				this.setupHover(this.descriptionNode.element, options?.descriptionTitle);
 			} else {
 				this.descriptionNode.textContent = description || '';
-				this.descriptionNode.title = options?.descriptionTitle || '';
+				this.setupHover(this.descriptionNode.element, options?.descriptionTitle || '');
 				this.descriptionNode.empty = !description;
 			}
 		}
 	}
+
+	private setupHover(htmlElement: HTMLElement, tooltip: string | Promise<IMarkdownString | string | undefined> | undefined): void {
+		if (!tooltip) {
+			htmlElement.removeAttribute('title');
+			return;
+		}
+
+		if (!this.hoverDelegate) {
+			htmlElement.title = isString(tooltip) ? tooltip : '';
+			return;
+		}
+
+		htmlElement.removeAttribute('title');
+		const hoverDelegate = this.hoverDelegate;
+		// Testing has indicated that on Windows and Linux 500 ms matches the native hovers most closely.
+		// On Mac, the delay is 1500.
+		const hoverDelay = isMacintosh ? 1500 : 500;
+		let hoverOptions: IHoverDelegateOptions | undefined;
+		let mouseX: number | undefined;
+		function mouseOver(this: HTMLElement, e: MouseEvent): any {
+			let isHovering = true;
+			function mouseMove(this: HTMLElement, e: MouseEvent): any {
+				mouseX = e.x;
+			}
+			function mouseLeave(this: HTMLElement, e: MouseEvent): any {
+				isHovering = false;
+			}
+			this.addEventListener(dom.EventType.MOUSE_LEAVE, mouseLeave, { passive: true });
+			this.addEventListener(dom.EventType.MOUSE_MOVE, mouseMove, { passive: true });
+			setTimeout(async () => {
+				if (isHovering && tooltip) {
+					// Re-use the already computed hover options if they exist.
+					if (!hoverOptions) {
+						const target: IHoverDelegateTarget = {
+							targetElements: [this],
+							dispose: () => { }
+						};
+						const resolvedTooltip = await tooltip;
+						if (resolvedTooltip) {
+							hoverOptions = { text: resolvedTooltip, target, anchorPosition: AnchorPosition.BELOW };
+						}
+					}
+					if (hoverOptions) {
+						if (mouseX !== undefined) {
+							(<IHoverDelegateTarget>hoverOptions.target).x = mouseX + 10;
+						}
+						hoverDelegate.showHover(hoverOptions);
+					}
+				}
+				this.removeEventListener(dom.EventType.MOUSE_MOVE, mouseMove);
+				this.removeEventListener(dom.EventType.MOUSE_LEAVE, mouseLeave);
+			}, hoverDelay);
+		}
+		htmlElement.addEventListener(dom.EventType.MOUSE_OVER, mouseOver, { passive: true });
+		this._register({
+			dispose: () => {
```

Comment:
This dispose is only being called when the entire label is disposed, but we seem to install the hover over and over everytime the label is updated? Do we not need to dispose previous hover handlers or handle that case properly?